### ISSUE TITLE
After updateing from 3.3 to 3.4 Object IDs are again names and not mac addresses

### ISSUE THREAD

**Issue by MatthiasBosch** [iobroker-community-adapters/ioBroker.chromecast#274]
### What happened?

When I update from version 3.3 to 3.4 Object IDs are again names and not mac addresses.
I can reproduce this behaviour.
Is this a planned switch back or a bug?

### Adapter Version

3.4

### JS-Controller Version

5.0.19

### Node JS Version

18.20.1

### NPM Version

10.5.0

### Operating System

Debian

**Comment by mcm1957** [comment]
If there is a name specified the name will be used again.
If no name is specified the mac address will be used. This was required to fix] https://github.com/iobroker-community-adapters/ioBroker.chromecast/issues/154 and https://github.com/iobroker-community-adapters/ioBroker.chromecast/issues/160. 

Could have been stated at releasenotes - sorry.

@neopholus 
Please clearify if required.

**Comment by Brainbug01** [comment]
Ja so ohne Informationen in den Changelogs ist das...
Nach dem Update auf 3.4 muss man die Sayit Instanz neu konfigurieren.

**Comment by raintonr** [comment]
> If no name is specified the mac address will be used.

This is not true. I have just upgraded to 3.4 which has broken my installation. Even when removing all adapter objects and letting discovery take place, players came back with names, not MAC addresses.

> Could have been stated at releasenotes - sorry.

This is a breaking change and should have been a major version update and noted as such. @mcm1957 please remember for the future.

**Comment by raintonr** [comment]
As an FYI. Rollback to 3.3.0 brings back object IDs as MAC addresses.

**Comment by raintonr** [comment]
@neopholus @mcm1957 ... this is the problem: https://github.com/iobroker-community-adapters/ioBroker.chromecast/commit/ef8584b5d2297293b0daee78cb8dbc26f3c38598 (well - that an subsequent changes which demote MAC address lookup which is then never attempted).

The scanner always returns a name (as defined in the Google Home app), and thus MAC address is never used.

I believe what you meant to do was look in the adapter configuration to see if a name was defined for a particular device (based on IP address I assume), and only in that case use the name defined in the adapter configuration (NOT the name provided by scanner from Google Home). Is that correct?

**Comment by neopholus** [comment]
> If there is a name specified the name will be used again. If no name is specified the mac address will be used. This was required to fix] #154 and #160 as mac addresses from other subnets cannot be retrieved automatically.
> 
> Could have been stated at releasenotes - sorry.
> 
> @neopholus Please clearify if required.

That was the intention, yes. 
As a quick workaround you could try to add the MAC address as a name in the configuration.

**Comment by neopholus** [comment]
> @neopholus @mcm1957 ... this is the problem: [ef8584b](https://github.com/iobroker-community-adapters/ioBroker.chromecast/commit/ef8584b5d2297293b0daee78cb8dbc26f3c38598) (well - that an subsequent changes which demote MAC address lookup which is then never attempted).
> 
> The scanner always returns a name (as defined in the Google Home app), and thus MAC address is never used.
> 
> I believe what you meant to do was look in the adapter configuration to see if a name was defined for a particular device (based on IP address I assume), and only in that case use the name defined in the adapter configuration (NOT the name provided by scanner from Google Home). Is that correct?

Hi @raintonr ,

the intention was to check if a name is defined in the adapter config and, if there is no name set for the current device, to try to get the MAC address and use it as name, correct. 
I looked at the code before the breaking change where MAC addresses have been used as ID and after the change. I had the intention, that this code

```
                    // Check if a name is defined in configuration
                    this.id = this._connection.name.replace(/[.\s]+/g, '_')
```

is referring to the name defined in the configuration. 

Sorry, if that was wrong,

**Comment by raintonr** [comment]
Well, it doesn't work like that for me.

No names are defined in adapter config but object IDs are not MAC address derived as expected. Instead they are taken from Google Home device names. 

Would you please fix this.

**Comment by neopholus** [comment]
@raintonr , @MatthiasBosch: I assume, your adapter config tab "devices" is empty, i.e., there are non only no names defined, but also no IP adresses, ports, and type? Please confirm.

Otherwise, could you please post a screenshot of the "devices" tab? Thanks!

**Comment by raintonr** [comment]
> @raintonr: I assume, your adapter config tab "devices" is empty...

Yes, of course.

**Comment by raintonr** [comment]
BTW, this is also horrible: https://github.com/iobroker-community-adapters/ioBroker.chromecast/commit/bf79d0630a3e49c369ddaabf7a51a2e1663507f8

```
// if somebody entered a MAC address, handle it like the result from getMac so that they look the same, i.e., remove colons
```

You should not blindly remove all colons. If the intent is to remove them from a MAC address, then check that the string looks like a MAC address with colons before removing them.

**Comment by neopholus** [comment]
> BTW, this is also horrible: [bf79d06](https://github.com/iobroker-community-adapters/ioBroker.chromecast/commit/bf79d0630a3e49c369ddaabf7a51a2e1663507f8)
> 
> ```
> // if somebody entered a MAC address, handle it like the result from getMac so that they look the same, i.e., remove colons
> ```
> 
> You should not blindly remove all colons. If the intent is to remove them from a MAC address, then check that the string looks like a MAC address with colons before removing them.

Comment is not good, I agree. But colons in state names are known to cause trouble, and they are part of FORBIDDEN_CHARACTERS in the meantime, see [https://github.com/ioBroker/ioBroker.js-controller/blob/4020943e2dc20d89672ab505a495384c62869987/packages/common/src/lib/common/tools.ts#L137](url), if I interpret the regexp correctly. Will test and then remove the comment and the next like as all forbidden characters will be removed anyhow.

Ok for you?

**Comment by neopholus** [comment]
> Well, it doesn't work like that for me.
> 
> No names are defined in adapter config but object IDs are not MAC address derived as expected. Instead they are taken from Google Home device names.
> 
> Would you please fix this.

Found a solution and do local testing now.

**Comment by neopholus** [comment]
@mcm1957 I implemented a fix for this issue and hope, that now everybody is happy. Can you please check and merge?

- Now it is tracked whether a device comes from the configuration or from the auto discovery.
- Devices from the configuration are created using the name from the configuration as it was done before 3.0.3 introduced MAC addresses as names for devices that has been auto-discovered. Unfortunately, devices added to the device tab of the configuration broke with that, as also there MAC addresses would have been used as name and not the configured name. For devices in different subnets, the getMAC failed, of course.
- Devices from the auto discovery are using MAC address as name, as introduced with 3.0.3.

@raintonr: Is this ok for you as well?

**Comment by neopholus** [comment]
Of course :-) I broke it, I fix it.

**Comment by neopholus** [comment]
@mcm1957 Sorry if this is a stupid question, but why did you reopen the issue?

**Comment by mcm1957** [comment]
Do not see that it was closed.
And ss far as I know its not fixed at any release. Only PR exists.raintonr doesnot Plan to Review at this time ....

**Comment by neopholus** [comment]
Sorry, I misinterpreted this:
![grafik](https://github.com/user-attachments/assets/d81801e2-8dfb-4192-9f80-43c8d022064b)

I think I understood it now.

**Comment by raintonr** [comment]
Any update on this? I upgraded to controller 6.x and now the old Chromecast adapter (the one without this bug) is incompatible... which makes the need to fix this more urgent.

**Comment by mcm1957** [comment]
PR #274 exists since August. Fell free to comment on it or test it.

**Comment by mcm1957** [comment]
fixed with 4.0.0
